source/linux/epoll_event_loop.c

```diff
@@ -394,22 +402,45 @@ static void s_on_tasks_to_schedule(
 
     struct epoll_loop *epoll_loop = event_loop->impl_data;
     if (events & AWS_IO_EVENT_TYPE_READABLE) {
-        uint64_t count_ignore = 0;
+        epoll_loop->should_process_task_pre_queue = true;
+    }
+}
 
-        aws_mutex_lock(&epoll_loop->task_pre_queue_mutex);
+static void s_process_task_pre_queue(struct aws_event_loop *event_loop) {
+    struct epoll_loop *epoll_loop = event_loop->impl_data;
 
-        /* several tasks could theoretically have been written (though this should never happen), make sure we drain the
-         * eventfd/pipe. */
-        while (read(epoll_loop->read_task_handle.data.fd, &count_ignore, sizeof(count_ignore)) > -1) {
-        }
+    if (!epoll_loop->should_process_task_pre_queue) {
+        return;
+    }
+
+    epoll_loop->should_process_task_pre_queue = false;
+
+    struct aws_linked_list task_pre_queue;
+    aws_linked_list_init(&task_pre_queue);
+
+    uint64_t count_ignore = 0;
 
-        while (!aws_linked_list_empty(&epoll_loop->task_pre_queue)) {
-            struct aws_linked_list_node *node = aws_linked_list_pop_front(&epoll_loop->task_pre_queue);
-            struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+    aws_mutex_lock(&epoll_loop->task_pre_queue_mutex);
+
+    /* several tasks could theoretically have been written (though this should never happen), make sure we drain the
+     * eventfd/pipe. */
+    while (read(epoll_loop->read_task_handle.data.fd, &count_ignore, sizeof(count_ignore)) > -1) {
+    }
+
+    aws_linked_list_swap_contents(&epoll_loop->task_pre_queue, &task_pre_queue);
+
+    aws_mutex_unlock(&epoll_loop->task_pre_queue_mutex);
+
+    while (!aws_linked_list_empty(&epoll_loop->task_pre_queue)) {
```

Comment:
I'm pretty sure this is a copy pasta error, and you meant:

while (!aws_linked_list_empty(&task_pre_queue)) {

etc.... I've already fixed it, pushing it up.